manager/container.go

```diff
@@ -572,9 +580,123 @@ func (c *containerData) updateStats() error {
+		if err != nil {
+			return fmt.Errorf("error while getting gpu uuid: %v", err)
+		}
+		memoryTotal, memoryUsed, err := device.MemoryInfo()
+		if err != nil {
+			return fmt.Errorf("error while getting gpu memory info: %v", err)
+		}
+		utilizationGPU, err := device.AverageGPUUtilization(c.housekeepingInterval)
+		if err != nil {
+			return fmt.Errorf("error while getting gpu utilization: %v", err)
+		}
+
+		stats.Accelerators = append(stats.Accelerators, info.AcceleratorStats{
+			Make:        "nvidia",
+			Model:       model,
+			ID:          uuid,
+			MemoryTotal: memoryTotal,
+			MemoryUsed:  memoryUsed,
+			DutyCycle:   uint64(utilizationGPU),
+		})
+	}
+	return nil
+}
+
+// getNvidiaMinorNumbers returns a list of minor numbers corresponding to NVIDIA GPU
+// devices that the container is allowed to access according to its devices cgroup
+func (c *containerData) getNvidiaMinorNumbers() []int {
+	if c.nvidiaMinorNumbers == nil {
+		var err error
+		c.nvidiaMinorNumbers, err = c.parseDevicesCgroup()
+		if err != nil {
+			glog.Infof("GPU metrics may be unavailable/incomplete for container %q: failed to parse devices cgroup: %v", c.info.Name, err)
+		}
+	}
+	return c.nvidiaMinorNumbers
+}
+
+// parseDevicesCgroup parses the devices cgroup devices.list file for the container
+// and returns a list of minor numbers corresponding to NVIDIA GPU devices that the
+// container is allowed to access. In cases where the container has access to all
+// devices or all NVIDIA devices but the devices are not enumerated separately in
+// the devices.list file, we return an empty list.
+func (c *containerData) parseDevicesCgroup() ([]int, error) {
+	// Always return a non-nil slice
+	nvidiaMinorNumbers := []int{}
+
+	devicesCgroupPath, err := c.handler.GetCgroupPath("devices")
+	if err != nil {
+		return nvidiaMinorNumbers, fmt.Errorf("error getting devices cgroup path: %v", err)
+	}
+
+	devicesList := filepath.Join(devicesCgroupPath, "devices.list")
+	f, err := os.Open(devicesList)
+	if err != nil {
+		return nvidiaMinorNumbers, fmt.Errorf("error while opening devices cgroup file %q: %v", devicesList, err)
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+
+	// See https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt for the file format
+	for s.Scan() {
+		text := s.Text()
+
+		fields := strings.Fields(text)
+		if len(fields) != 3 {
+			return nvidiaMinorNumbers, fmt.Errorf("invalid devices cgroup entry %q: must contain three whitespace-separated fields", text)
+		}
+
+		// Split the second field to find out major:minor numbers
+		majorMinor := strings.Split(fields[1], ":")
+		if len(majorMinor) != 2 {
+			return nvidiaMinorNumbers, fmt.Errorf("invalid devices cgroup entry %q: second field should have one colon", text)
+		}
+
+		// NVIDIA graphics devices are character devices with major number 195.
+		// https://github.com/torvalds/linux/blob/v4.13/Documentation/admin-guide/devices.txt#L2583
+		if fields[0] == "c" && majorMinor[0] == "195" {
```

Comment:
@flx42 Are these APIs reliable enough for at-least a year?

Reply:
It's not an API, it's an implementation detail that happens to work with Docker.
There are other ways to block GPUs besides groups, for instance you can bind mount the devices and then block `mknod`, for instance when using user namespaces.

That being said, for the Docker use case, I'm not aware of any upcoming breaking change, but my vision doesn't go up to one year forward.